src/QsCompiler/CommandLineTool/CompilationTracker.cs

```diff
@@ -0,0 +1,281 @@
+        private enum WarningType
+        {
+            ProcessAlreadyExists,
+            ProcessDoesNotExist,
+            ProcessAlreadyEnded
+        }
+
+        /// <summary>
+        /// Represents a warning detected while tracking compilation events.
+        /// </summary>
+        private class Warning
+        {
+            public readonly WarningType Type;
+            public readonly DateTime UtcDateTime;
+            public readonly string Key;
+
+            public Warning(WarningType type, string key)
+            {
+                UtcDateTime = DateTime.UtcNow;
+                Type = type;
+                Key = key;
+            }
+        }
+
+        // Private members.
+
+        /// <summary>
+        /// Defines a handler for a type of compilation task event.
+        /// </summary>
+        private delegate void CompilationTaskEventTypeHandler(CompilationLoader.CompilationTaskEventArgs eventArgs);
+
+        /// <summary>
+        /// Contains a handler that takes care of each type of task event.
+        /// </summary>
+        private static readonly IDictionary<CompilationLoader.CompilationTaskEventType, CompilationTaskEventTypeHandler> CompilationEventTypeHandlers = new Dictionary<CompilationLoader.CompilationTaskEventType, CompilationTaskEventTypeHandler>
+        {
+            { CompilationLoader.CompilationTaskEventType.Start, CompilationEventStartHandler },
+            { CompilationLoader.CompilationTaskEventType.End, CompilationEventEndHandler }
+        };
+
+        /// <summary>
+        /// Contains the compilation tasks tracked through the handled events.
+        /// </summary>
+        private static readonly IDictionary<string, CompilationTask> CompilationTasks = new Dictionary<string, CompilationTask>();
+        /// <summary>
+        /// Contains the warnings generated while handling the compiler tasks events.
+        /// </summary>
+        private static readonly IList<Warning> Warnings = new List<Warning>();
+
+        // Private methods.
+
+        /// <summary>
+        /// Creates a hierarchical structure that contains the compilation tasks.
+        /// </summary>
+        private static IList<CompilationTaskNode> BuildCompilationTasksHierarchy()
+        {
+            var compilationTasksForest = new List<CompilationTaskNode>();
+            var toFindChildrenNodes = new Queue<CompilationTaskNode>();
+
+            // First add the roots (top-level tasks) of all trees to the forest.
+
+            foreach (var entry in CompilationTasks)
+            {
+                if (entry.Value.ParentName == null)
+                {
+                    var node = new CompilationTaskNode(entry.Value);
+                    compilationTasksForest.Add(node);
+                    toFindChildrenNodes.Enqueue(node);
+                }
+            }
+
+            // Iterate through the tasks until all of them have been added to the hierarchy.
+
+            while (toFindChildrenNodes.Count > 0)
```

Comment:
FYI, there is also TryDequeue, though it seems like Count is also O(1) plus it really doesn't seem relevant which one you use in practice in this case.

Reply:
C# Queue class does not seem to have a TryDequeue method, only the ConcurrentQueue which there is no need for in this case. I will keep using Count in this case :)